// bespoken-tools/lib/client/bst-alexa.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var interaction_model_1 = require("../alexa/interaction-model");
var alexa_1 = require("../alexa/alexa");
var global_1 = require("../core/global");
var service_request_1 = require("../alexa/service-request");
var BSTAlexaEvents = (function () {
    function BSTAlexaEvents() {
    }
    BSTAlexaEvents.AudioPlayerPlaybackFinished = "AudioPlayer.PlaybackFinished";
    BSTAlexaEvents.AudioPlayerPlaybackNearlyFinished = "AudioPlayer.PlaybackNearlyFinished";
    BSTAlexaEvents.AudioPlayerPlaybackStarted = "AudioPlayer.PlaybackStarted";
    BSTAlexaEvents.AudioPlayerPlaybackStopped = "AudioPlayer.PlaybackStopped";
    BSTAlexaEvents.Response = "response";
    return BSTAlexaEvents;
}());
exports.BSTAlexaEvents = BSTAlexaEvents;
var BSTAlexa = (function () {
    function BSTAlexa(skillURL, intentSchemaFile, sampleUtterancesFile, applicationID) {
        this.skillURL = skillURL;
        this.intentSchemaFile = intentSchemaFile;
        this.sampleUtterancesFile = sampleUtterancesFile;
        this.applicationID = applicationID;
        this._alexa = null;
        if (!this.intentSchemaFile) {
            this.intentSchemaFile = BSTAlexa.DefaultIntentSchemaLocation;
        }
        if (!this.sampleUtterancesFile) {
            this.sampleUtterancesFile = BSTAlexa.DefaultSampleUtterancesLocation;
        }
        this._alexa = new alexa_1.Alexa();
        if (global_1.Global.config()) {
            if (this.applicationID) {
                global_1.Global.config().updateApplicationID(this.applicationID);
            }
            else {
                this.applicationID = global_1.Global.config().applicationID();
            }
        }
    }
    BSTAlexa.prototype.start = function (ready) {
        var self = this;
        interaction_model_1.InteractionModel.fromFiles(this.intentSchemaFile, this.sampleUtterancesFile, function (model, error) {
            if (error !== undefined && error !== null) {
                ready(error);
            }
            else {
                self._alexa.startSession(self.skillURL, model, true, self.applicationID);
                ready();
            }
        });
    };
    BSTAlexa.prototype.context = function () {
        return this._alexa.context();
    };
    BSTAlexa.prototype.on = function (eventType, callback) {
        if (eventType.startsWith("AudioPlayer")) {
            if (!BSTAlexa.validateAudioEventType(eventType)) {
                throw Error("No event type: " + eventType + " is defined");
            }
            if (this._alexa.context().audioPlayerEnabled()) {
                this._alexa.context().audioPlayer().on(eventType, callback);
            }
        }
        else if (eventType === BSTAlexaEvents.Response) {
            this._alexa.on(alexa_1.AlexaEvent.SkillResponse, callback);
        }
        else {
            throw Error("No event type: " + eventType + " is defined");
        }
        return this;
    };
    BSTAlexa.prototype.once = function (eventType, callback) {
        if (eventType.startsWith("AudioPlayer")) {
            if (!BSTAlexa.validateAudioEventType(eventType)) {
                throw Error("No event type: " + eventType + " is defined");
            }
            if (this._alexa.context().audioPlayerEnabled()) {
                this._alexa.context().audioPlayer().once(eventType, callback);
            }
        }
        else if (eventType === BSTAlexaEvents.Response) {
            this._alexa.once(alexa_1.AlexaEvent.SkillResponse, callback);
        }
        else {
            throw Error("No event type: " + eventType + " is defined");
        }
        return this;
    };
    BSTAlexa.prototype.spoken = function (phrase, callback) {
        this._alexa.spoken(phrase, function (error, response, request) {
            if (callback !== undefined && callback !== null) {
                callback(error, response, request);
            }
        });
        return this;
    };
    BSTAlexa.prototype.intended = function (intentName, slots, callback) {
        this._alexa.intended(intentName, slots, function (error, response, request) {
            if (callback !== undefined && callback !== null) {
                callback(error, response, request);
            }
        });
        return this;
    };
    BSTAlexa.prototype.launched = function (callback) {
        this._alexa.launched(callback);
        return this;
    };
    BSTAlexa.prototype.sessionEnded = function (sessionEndedReason, callback) {
        var sessionEndedEnum = service_request_1.SessionEndedReason[sessionEndedReason];
        this._alexa.sessionEnded(sessionEndedEnum, null, callback);
        return this;
    };
    BSTAlexa.prototype.playbackFinished = function (callback) {
        if (this._alexa.context().audioPlayerEnabled()) {
            if (this._alexa.context().audioPlayer().isPlaying()) {
                this._alexa.context().audioPlayer().playbackFinished(callback);
            }
        }
        return this;
    };
    BSTAlexa.prototype.playbackNearlyFinished = function (callback) {
        if (this._alexa.context().audioPlayerEnabled()) {
            if (this._alexa.context().audioPlayer().isPlaying()) {
                this._alexa.context().audioPlayer().playbackNearlyFinished(callback);
            }
        }
        return this;
    };
    BSTAlexa.prototype.playbackStopped = function (callback) {
        if (this._alexa.context().audioPlayerEnabled()) {
            if (this._alexa.context().audioPlayer().isPlaying()) {
                this._alexa.context().audioPlayer().playbackStopped(callback);
            }
        }
        return this;
    };
    BSTAlexa.prototype.playbackOffset = function (offsetInMilliseconds) {
        this._alexa.context().audioPlayer().playbackOffset(offsetInMilliseconds);
        return this;
    };
    BSTAlexa.prototype.stop = function (onStop) {
        this._alexa.stop(onStop);
    };
    BSTAlexa.validateAudioEventType = function (eventType) {
        var match = false;
        for (var _i = 0, _a = BSTAlexa.AudioPlayerEvents; _i < _a.length; _i++) {
            var e = _a[_i];
            if (eventType === e) {
                match = true;
                break;
            }
        }
        return match;
    };
    BSTAlexa.AudioPlayerEvents = [BSTAlexaEvents.AudioPlayerPlaybackFinished,
        BSTAlexaEvents.AudioPlayerPlaybackNearlyFinished,
        BSTAlexaEvents.AudioPlayerPlaybackStarted,
        BSTAlexaEvents.AudioPlayerPlaybackStopped];
    BSTAlexa.DefaultIntentSchemaLocation = "speechAssets/IntentSchema.json";
    BSTAlexa.DefaultSampleUtterancesLocation = "speechAssets/SampleUtterances.txt";
    return BSTAlexa;
}());
exports.BSTAlexa = BSTAlexa;
